File: src/test/java/net/sf/rails/game/state/IntegerStateTest.java
package net.sf.rails.game.state;

import static org.fest.assertions.api.Assertions.assertThat;
import static org.junit.Assert.*;

import net.sf.rails.game.state.IntegerState;
import net.sf.rails.game.state.Root;

import org.junit.Before;
import org.junit.Test;

public class IntegerStateTest {

    private final static String DEFAULT_ID = "Default";
    private final static String INIT_ID = "Init"; 
    private final static int INIT = 10;
    private final static int OTHER = -5;

    private Root root;
    
    private IntegerState stateDefault;
    private IntegerState stateInit;
    

    @Before
    public void setUp() {
        root = StateTestUtils.setUpRoot();
        
        
        stateDefault = IntegerState.create(root, DEFAULT_ID);
        stateInit = IntegerState.create(root, INIT_ID, INIT);
    }
    
    @Test
    public void testValue() {
        assertEquals(stateDefault.value(), 0);
        assertEquals(stateInit.value(), INIT);
    }

    @Test
    public void testSet() {
        stateDefault.set(OTHER);
        assertEquals(stateDefault.value(), OTHER);
        stateInit.set(0);
        assertEquals(stateInit.value(), 0);
        
    }

    @Test
    public void testAdd() {
        stateDefault.add(OTHER);
        assertEquals(stateDefault.value(), OTHER);
        stateInit.add(OTHER);
        assertEquals(stateInit.value(), INIT + OTHER);
    }
    
    
    @Test
    public void testSetSameIgnored() {
        stateDefault.set(0);
        stateInit.set((INIT));
        StateTestUtils.close(root);
        assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).doesNotContain(stateDefault, stateInit);
    }

    @Test
    public void testUndoRedo() {
        
        stateDefault.set(INIT);
        stateDefault.add(OTHER);
        
        stateInit.add(OTHER);
        stateInit.set(0);
        StateTestUtils.close(root);
        
        assertEquals(stateDefault.value(), INIT+OTHER);
        assertEquals(stateInit.value(), 0);

        StateTestUtils.undo(root);
        assertEquals(stateDefault.value(), 0);
        assertEquals(stateInit.value(), INIT);

        StateTestUtils.redo(root);
        assertEquals(stateDefault.value(), INIT+OTHER);
        assertEquals(stateInit.value(), 0);
    }

}

File: dotnet/GameStateTest/IntegerStateTest.cs
using GameLib.Net.Game.State;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameStateTest
{
    [TestClass]
    public class IntegerStateTest
    {
        private const string DEFAULT_ID = "Default";
        private const string INIT_ID = "Init";
        private const int INIT = 10;
        private const int OTHER = -5;

        private Root root;

        private IntegerState stateDefault;
        private IntegerState stateInit;


        [TestInitialize]
        public void SetUp()
        {
            root = StateTestUtils.SetUpRoot();


            stateDefault = IntegerState.Create(root, DEFAULT_ID);
            stateInit = IntegerState.Create(root, INIT_ID, INIT);
        }

        [TestMethod]
        public void TestValue()
        {
            Assert.AreEqual(stateDefault.Value, 0);
            Assert.AreEqual(stateInit.Value, INIT);
        }

        [TestMethod]
        public void TestSet()
        {
            stateDefault.Set(OTHER);
            Assert.AreEqual(stateDefault.Value, OTHER);
            stateInit.Set(0);
            Assert.AreEqual(stateInit.Value, 0);
        }

        [TestMethod]
        public void TestAdd()
        {
            stateDefault.Add(OTHER);
            Assert.AreEqual(stateDefault.Value, OTHER);
            stateInit.Add(OTHER);
            Assert.AreEqual(stateInit.Value, INIT + OTHER);
        }


        [TestMethod]
        public void TestSetSameIgnored()
        {
            stateDefault.Set(0);
            stateInit.Set((INIT));
            StateTestUtils.Close(root);
            var a = StateTestUtils.GetPreviousChangeSet(root).GetStates();
            Assert.IsTrue(!(a.Contains(stateDefault) || a.Contains(stateInit)));
            //assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).doesNotContain(stateDefault, stateInit);
        }

        [TestMethod]
        public void TestUndoRedo()
        {

            stateDefault.Set(INIT);
            stateDefault.Add(OTHER);

            stateInit.Add(OTHER);
            stateInit.Set(0);
            StateTestUtils.Close(root);

            Assert.AreEqual(stateDefault.Value, INIT + OTHER);
            Assert.AreEqual(stateInit.Value, 0);

            StateTestUtils.Undo(root);
            Assert.AreEqual(stateDefault.Value, 0);
            Assert.AreEqual(stateInit.Value, INIT);

            StateTestUtils.Redo(root);
            Assert.AreEqual(stateDefault.Value, INIT + OTHER);
            Assert.AreEqual(stateInit.Value, 0);
        }
    }
}
